File: Leetcode/1046 Last Stone Weight.py
class Solution:
    def lastStoneWeight(self, stones):
        # used -ve to create max heap
        for i in range(len(stones)):
            stones[i] *= -1
        heapify(stones)
        while len(stones) > 1:
            y = -heappop(stones)
            x = -heappop(stones)
            if x != y:
                heappush(stones, -(y - x))
        return -stones[0] if stones else 0

File: Leetcode/1046 Last Stone Weight.java
class Solution {
    public int lastStoneWeight(int[] stones) {
        var heap = new PriorityQueue<Integer>(Comparator.reverseOrder());
        for (var stone : stones) {
            heap.add(stone);
        }
        while (heap.size() > 1) {
            int y = heap.poll(), x = heap.poll();
            if (x != y) {
                heap.add(y - x);
            }
        }
        return heap.isEmpty()? 0 : heap.peek(); 
    }
}
